src/base/path.h

```diff
@@ -40,9 +40,23 @@ class Path final
     Path() = default;
 
     explicit Path(const QString &pathStr);
-    explicit Path(const char pathStr[]);
     explicit Path(const std::string &pathStr);
 
+    template <std::size_t N>
+    explicit Path(const char (&pathStr)[N])
```

Comment:
Is it worth it to defend against the almost impossible case when some developer adds non-ascii string literal to the code?

Reply:
Having compile-time checks enforcing the correct usage shows the intention and is worth it IMO.

Reply:
>Having compile-time checks

But it will produce multiple similar constructors just for compile-time checking. Are you sure they will be inlined and not pollute target binary?
>enforcing the correct usage

We actively use QLatin1String for the same purpose and we never had issues with incorrect usage, have we? 
>However this commit is dropped due to technical difficulties.

Maybe just add CI checking?

Reply:
Or maybe use Unicode literals instead? We can drop this constructor and add `Path(QStringView)`.
Of course we still can step back to use `Path(QLatin1String("some/path"))`.

Reply:
>But it will produce multiple similar constructors just for compile-time checking.

What are you talking about? There is only that constructor template (which is reverted now).

>We actively use QLatin1String for the same purpose and we never had issues with incorrect usage, have we?

It is lame IMO that compiler (and Qt) couldn't do better. Just think outside of a specific language construct, how hard is it to check (at compile time) some string literal and provide different code path (or even compile warnings) for it?

As said before, I dropped the change and I am not pursuing this goal now.

Reply:
>Or maybe use Unicode literals instead? We can drop this constructor and add Path(QStringView).

(Outside of this PR) I think this make sense and this idea could be applied through out our code base.

Reply:
> > But it will produce multiple similar constructors just for compile-time checking.
> 
> What are you talking about? There is only that constructor template (which is reverted now).

Compiler instantiates template for each unique set of template arguments, doesn't it?

Reply:
> > Or maybe use Unicode literals instead? We can drop this constructor and add Path(QStringView).
> 
> (Outside of this PR) I think this make sense and this idea could be applied throughout our code base.

I wanted to do this initially, but I didn't want to get involved with complaints about increasing the size of the binary (by a couple of kilobytes). I can't remember how you feel about it, but someone always complains...

Reply:
> > Having compile-time checks
> 
> But it will produce multiple similar constructors just for compile-time checking. Are you sure they will be inlined and not pollute target binary?

And why would it bloat the binary when the body of it is only compile-time checks? I don't think that would happen.

Reply:
>I wanted to do this initially, but I didn't want to get involved with complaints about increasing the size of the binary (by a couple of kilobytes). 

Note that QString is always taking a copy of a string literal and `QStringLiteral` is the only way to avoid that. 
QStringView can take utf-16 literals but not utf-8. 

>I can't remember how you feel about it, but someone always complains...

🤷‍♂️ maybe you took the wrong approach at that time?

Reply:
> Note that QString is always taking a copy of a string literal and `QStringLiteral` is the only way to avoid that.

Yes. But if we use utf16 literals instead of any kind of 8-bit ones (include utf8) we always have larger binary. `QStringLiteral` just keep us from runtime overhead (by initializing QString without making copy of literal).

Reply:
>Yes. But if we use utf16 literals instead of any kind of 8-bit ones (include utf8) we always have larger binary. 

Right, I was implying that using utf16 literals is only beneficial in the case of QStringView.
For QString, using utf16 literal only saves some keyboard typing (compared to utf8 literals + QString::fromLatin1()) and it took twice the size on binary. So I don't think there is any advantage of it.

Reply:
> For QString, using utf16 literal only saves some keyboard typing (compared to utf8 literals + QString::fromLatin1()) and it took twice the size on binary. So I don't think there is any advantage of it.

The advantage is avoiding transcoding. But it really makes sense in case of frequently used string constants.

Reply:
>The advantage is avoiding transcoding.

For ascii literals, QString::fromLatin1() can do it and it is efficient right?
For utf8 literals, maybe I'm wrong but I don't think there are many instances of it in qbt code base.
For utf16 literals, QString::fromUtf16() is deemed slow from Qt docs: https://doc.qt.io/qt-5/qstring.html#fromUtf16

Reply:
> For ascii literals, QString::fromLatin1() can do it and it is efficient right?

Right, it uses some advanced CPU features like SSE/AVX but it isn't so efficient as just copying existing utf16 data.
In any case, it all comes down to the age-old dilemma of optimization "size vs. speed".
>For utf16 literals, `QString::fromUtf16()` is deemed slow from Qt docs

`QString::fromUtf16()` seems to be useful when reading some UTF-16 encoded file since it checks for BOM.
To initialize QString with from utf16 literals we can use either `QStringLiteral`, `QStringView::toString()` or `QString(const QChar *)`.

Reply:
>To initialize QString with from utf16 literals we can use either QStringLiteral, QStringView::toString() or QString(const QChar *).

The first 2 is clear to me: `QStringLiteral(u"abc")`, `QStringView(u"abc")::toString()`. But I'm not sure how to write for the third. `QString(u"abc")` gives compile errors.

Reply:
>But I'm not sure how to write for the third. QString(u"abc") gives compile errors.

I mean if we can find a way to do it then it would definitely be beneficial to use utf16 literals ubiquitously in the code base.

Reply:
I think we should port this to our code base for the sake of supporting Qt5: https://doc.qt.io/qt-6/qstring.html#operator-22-22_qs and start using utf16 literals.

Reply:
>Just think outside of a specific language construct, how hard is it to check (at compile time) some string literal and provide different code paths (or even compile warnings) for it?

The compiler has all the necessary information about string literals to allow you to do this. It remains only to provide this opportunity to users of the programming language.

Reply:
> I think we should port this to our code base for the sake of supporting Qt5: https://doc.qt.io/qt-6/qstring.html#operator-22-22_qs and start using utf16 literals.

I would try.
But you confuse me with your attitude to increasing the size of the binary. In some comments you speak as if you disapprove of it, in others in a different way...

Reply:
>But you confuse me with your attitude to increasing the size of the binary. 

Not sure what confuse you... as long as there are performance benefits, the size is (usually) not a problem.

Reply:
> Are you going to handle it? or I can put some effort into it.

I had other plans. Besides, it's your idea.